src/pudl/metadata/classes.py

```diff
@@ -661,59 +661,85 @@ def to_pyarrow(self) -> pa.Field:
-            if self.constraints.enum:
-                enum = [_format_for_sql(x) for x in self.constraints.enum]
-                checks.append(f"{name} IN ({', '.join(enum)})")
+        field_type = self.to_sql_dtype()
+        if dialect == "sqlite":
+            if check_types:
+                # Required with TYPEOF since TYPEOF(NULL) = 'null'
+                prefix = "" if self.constraints.required else f"{name} IS NULL OR "
+                # Field type
+                if self.type == "string":
+                    checks.append(f"{prefix}TYPEOF({name}) = 'text'")
+                elif self.type in ("integer", "year"):
+                    checks.append(f"{prefix}TYPEOF({name}) = 'integer'")
+                elif self.type == "number":
+                    checks.append(f"{prefix}TYPEOF({name}) = 'real'")
+                elif self.type == "boolean":
+                    # Just IN (0, 1) accepts floats equal to 0, 1 (0.0, 1.0)
+                    checks.append(
+                        f"{prefix}(TYPEOF({name}) = 'integer' AND {name} IN (0, 1))"
+                    )
+                elif self.type == "date":
+                    checks.append(f"{name} IS DATE({name})")
+                elif self.type == "datetime":
+                    checks.append(f"{name} IS DATETIME({name})")
+            if check_values:
+                # Field constraints
+                if self.constraints.min_length is not None:
+                    checks.append(f"LENGTH({name}) >= {self.constraints.min_length}")
+                if self.constraints.max_length is not None:
+                    checks.append(f"LENGTH({name}) <= {self.constraints.max_length}")
+                if self.constraints.minimum is not None:
+                    minimum = _format_for_sql(self.constraints.minimum)
+                    checks.append(f"{name} >= {minimum}")
+                if self.constraints.maximum is not None:
+                    maximum = _format_for_sql(self.constraints.maximum)
+                    checks.append(f"{name} <= {maximum}")
+                if self.constraints.pattern:
+                    pattern = _format_for_sql(self.constraints.pattern)
+                    checks.append(f"{name} REGEXP {pattern}")
+                if self.constraints.enum:
+                    enum = [_format_for_sql(x) for x in self.constraints.enum]
+                    checks.append(f"{name} IN ({', '.join(enum)})")
+        elif dialect == "duckdb":
+            if isinstance(field_type, sa.Enum):
+                # TODO: Not supporting enums with duckdb right now because
+                # code tables primary keys are VARCHARS but the tables that
+                # reference them have ENUM types. Duckdb throws an error
+                # because of the mismatched types. It is ok for now because
+                # the enums are checked in SQLite and the foreign key relationship
+                # acts as an enum constraint.
```

Comment:
Are the ENUM types in the tables that reference the coding tables over-specifying the values? Like are they doubly constrained, by both the ENUM and by the FK?  I saw some ENUMs like that creep in a while ago but it seemed like they probably shouldn't be there -- just having the FK relationship should be sufficient to ensure that all the values are valid.

Reply:
Yes, they are double-contained. This wasn't an issue with SQLite because the columns in both tables are TEXT types and we add a CHECK constraint to the enum columns to make sure all the values are valid. 

Our `Field.to_sql_dtype()` method creates a `sa.Enum` type. Maybe we can just make enums an `sa.String` type? Do you know if we have any enum columns that aren't validated by a foreign key constraint?

Reply:
Hmm. IIRC there's a big space savings for using ENUM in Parquet (maybe in DuckDB too?) in the big tables (where strings get dictionary encoded as ints). Maybe we just want to make both the PK and FK columns into ENUMs to satisfy DuckDB's (very reasonable) nitpickiness around dtypes?  Even though that'll be over-constrained in terms of limiting the allowed values? Not sure what the best approach is.

Reply:
Do you have ideas on how we can make the `code` field in the code tables an enum type using our metadata tooling? It feels tricky because the `code` field is shared among all the code tables, but they all have different enum values.

The `parquet_to_duckdb` script freezes after loading a few dozen tables. I removed the foreign key constraints from the database schema, and the data loaded just fine. I'm assuming the slowdown is happening because it's spending a bunch of time checking the sometimes 5 - 6 foreign key constraints on a table. I wonder if enabling the enums will help with this issue?

Reply:
Can we define the various different ENUM's for the `code` columns in all the different coding tables on a per-resource basis in the special cases at the bottom of the resource metadata files?